### PR TITLE
📨 Add a basic input abstraction

### PR DESCRIPTION
**Pull request by yufeih**
Depending on #5003 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5034)

### CODE REVIEW

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 1]
```diff
@@ -288,34 +276,32 @@ private string ApplyResolveAlias(string path, Document referencingFile)
             return docsetRelativePath;
         }
 
-        private Document TryResolveFile(Document referencingFile, string pathToDocset)
+        private Document TryResolveFile(Document declaringFile, string pathToDocset, bool lookupFallbackCommits)
```
Comment (live1206) [inline src/docfx/build/dependency/DependencyResolver.cs]:
```suggestion
        private Document TryResolveFile(Document referencingFile, string pathToDocset, bool lookupFallbackCommits)
```

File: src/docfx/lib/git/GitUtility.cs [hunk 2]
```diff
@@ -380,5 +371,33 @@ string GetSecretFromHeader(KeyValuePair<string, string> header)
                 return header.Value;
             }
         }
+
+        private class GitBlobStream : UnmanagedMemoryStream
+        {
+            private bool _disposed;
+            private IntPtr _repo;
```
Comment (live1206) [inline src/docfx/lib/git/GitUtility.cs]:
```suggestion
            private readonly IntPtr _repo;
```

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 3]
```diff
@@ -326,56 +312,23 @@ private Document TryResolveFile(Document referencingFile, string pathToDocset)
...
-                var (repo, pathToRepo, commits) = gitCommitProvider.GetCommitHistory(_fallbackDocset, pathToDocset);
-                if (repo != null)
+                // resolve from fallback docset git commit history
+                if (lookupFallbackCommits)
                 {
-                    var repoPath = PathUtility.NormalizeFolder(repo.Path);
-                    if (commits.Length > 1)
+                    var (repo, _, commits) = _gitCommitProvider.GetCommitHistory(_fallbackDocset, pathToDocset);
+
+                    // TODO: need to get the first commit that contains the data
```
Comment (live1206) [inline src/docfx/build/dependency/DependencyResolver.cs]: Could you elaborate a bit more about this TODO, does this mean `commits[1]` might not be the one containing data?

File: src/docfx/build/context/Input.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                ?? throw new InvalidOperationException($"Error reading '{file}'");
+
+            return new StreamReader(new MemoryStream(bytes, writable: false));
+        }
+
+        private (string basePath, string path, string commit) ResolveFilePath(FilePath file)
+        {
+            switch (file.Origin)
+            {
+                case FileOrigin.Default when file.Commit is null:
```
Comment (OsmondJiang) [inline src/docfx/build/context/Input.cs]: >case [](start = 16, length = 4)

From input abstraction view of point, it's better avoid to invloving too many business logic in Input.

Default, Dependency, Fallback, Template can all map to Physical, Git and Memory.

Considering that in the future we add new business dependency model, we don't want to change Input to support it.

Comment (yufeih) [reply src/docfx/build/context/Input.cs]: The design goal of `FilePath` is serializable and roamable (machine independent), thus it is designed this way. Using __physical/git/memory__ is machine dependent.

Comment (OsmondJiang) [reply src/docfx/build/context/Input.cs]: physical/git/memory is not always be machine dependent, and also, we should go details about the puporse of serializable/roadmable.

---
In reply to: [320084723](https://github.com/dotnet/docfx/pull/5034#discussion_r320084723) [](ancestors = 320084723)

File: src/docfx/build/context/Input.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            {
+                return false;
+            }
+
+            if (commit is null)
+            {
+                return File.Exists(Path.Combine(basePath, path));
+            }
+
+            throw new NotSupportedException($"{nameof(Exists)}: {file}");
```
Comment (OsmondJiang) [inline src/docfx/build/context/Input.cs]: TODO?

File: src/docfx/build/context/Input.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            switch (file.Origin)
+            {
+                case FileOrigin.Default when file.Commit is null:
+                    return (_docsetPath, file.Path, null);
+
+                case FileOrigin.Dependency when file.Commit is null:
+                    var (dependencyPath, _) = _restoreMap.GetGitRestorePath(_config.Dependencies[file.DependencyName], _docsetPath);
+                    return (dependencyPath, file.Path, null);
+
+                case FileOrigin.Fallback when file.Commit is null && _fallbackPath != null:
```
Comment (OsmondJiang) [inline src/docfx/build/context/Input.cs]: a little duplicated

File: src/docfx/build/context/Input.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            if (commit is null)
+            {
+                return File.Exists(Path.Combine(basePath, path));
+            }
+
+            throw new NotSupportedException($"{nameof(Exists)}: {file}");
+        }
+
+        /// <summary>
+        /// Try get the absolute path of the specified file if it exit physically on disk.
```
Comment (live1206) [inline src/docfx/build/context/Input.cs]:
```suggestion
        /// Try get the absolute path of the specified file if it exists physically on disk.
```

**Pull request merged**